### PR TITLE
Add `snapped` to integer vectors

### PR DESCRIPTION
**Pull request by joriskleiber**
This adds  `snapped` for integer vectors based on this [code](https://github.com/godot-rust/gdext/pull/721#discussion_r1615336780).

### CODE REVIEW

File: godot-core/src/builtin/vectors/vector_macros.rs [hunk 1]
```diff
@@ -457,6 +457,42 @@ macro_rules! impl_vector_fns {
...
+        /// component in `step`.
+        pub fn snapped(self, step: Self) -> Self {
+            #[inline]
+            fn snap_one(mut value: i32, step: i32) -> i32 {
+                if step != 0 {
+                    let a = value + step / 2;
+
+                    // manual implement `a.div_floor(step)` since Rust's native method is still unstable, as of 1.79.0
+                    let mut d = a / step;
+                    let r = a % step;
```
Comment (joriskleiber) [inline godot-core/src/builtin/vectors/vector_macros.rs:476]: This can't [panic](https://doc.rust-lang.org/std/primitive.i32.html#panics-25) because if `a` is ever equal to `i32::MIN`(`-2147483648`) and `step` is `-1`, `value` needs to be `-2147483647.5` which is imposible.

Comment (lilizoey) [reply godot-core/src/builtin/vectors/vector_macros.rs:476]: if `value` is `i32::MIN` and step is `-1` then `a` is `i32::MIN` since `-1 / 2 = 0` for `i32`.

File: godot-core/src/builtin/vectors/vector_macros.rs [hunk 2]
```diff
@@ -457,6 +457,42 @@ macro_rules! impl_vector_fns {
...
+/// Implements functions that are present only on integer vectors.
+macro_rules! impl_integer_vector_fns {
+    (
+        // Names of the components, for example `x, y`.
+        $($comp:ident),*
+    ) => {
+        /// A new vector with each component snapped to the closest multiple of the corresponding
+        /// component in `step`.
+        pub fn snapped(self, step: Self) -> Self {
+            #[inline]
```
Comment (Bromeon) [inline godot-core/src/builtin/vectors/vector_macros.rs:469]: `#[inline]` is mostly needed for cross-crate inlining (because Rust simply can't do it without that attribute), but I wouldn't use it for private/crate-local functions. The compiler should be able to judge whether to inline such functions itself.

Comment (Bromeon) [reply godot-core/src/builtin/vectors/vector_macros.rs:469]: Additionally, the `snap_one` method is now duplicated for all vectors. I think it would be better to declare it once manually outside the macro.

**Review by Bromeon — COMMENTED** [review]
Thank you!

Could you also add tests for zero steps? (Either one component 0, or the whole step vector).

File: godot-core/src/builtin/vectors/vector_macros.rs (hunk 2 above)
Comment (lilizoey) [reply godot-core/src/builtin/vectors/vector_macros.rs:469]: > `#[inline]` is mostly needed for cross-crate inlining (because Rust simply can't do it without that attribute)

Not entirely true anymore, rust will automatically inline small functions as of 1.75 https://github.com/rust-lang/rust/pull/116505

Comment (Bromeon) [reply godot-core/src/builtin/vectors/vector_macros.rs:469]: Yeah, this is however hardly useful in practice -- even if a function is "simple", there's a chance that refactoring makes it non-simple and causes pessimizations. As such, always writing `#[inline]` when inlining is allowed is still the better practice...

But it's a start, I hope this extended in the not too distant future 🙂

**Comment by joriskleiber** [comment]
I don't know if there is a better way to handle the possible division overflow, but I thinks its better to panic with a message then just letting it crash.

File: godot-core/src/builtin/vectors/vector_macros.rs [hunk 3]
```diff
@@ -457,6 +457,51 @@ macro_rules! impl_vector_fns {
...
+pub(super) fn snap_one(mut value: i32, step: i32) -> i32 {
+    assert!(
+        value != i32::MIN || step != -1,
+        "snapped() called on vector component i32::MIN with step component -1"
+    );
+
+    if step != 0 {
+        let a = value + step / 2;
+
+        // manual implement `a.div_floor(step)` since Rust's native method is still unstable, as of 1.79.0
```
Comment (Bromeon) [inline godot-core/src/builtin/vectors/vector_macros.rs:469]:
```suggestion
        // Manually implement `a.div_floor(step)` since Rust's native method is still unstable, as of 1.79.0.
```

File: godot-core/src/builtin/vectors/vector_macros.rs [hunk 4]
```diff
@@ -457,6 +457,51 @@ macro_rules! impl_vector_fns {
     }
 }
 
+pub(super) fn snap_one(mut value: i32, step: i32) -> i32 {
+    assert!(
+        value != i32::MIN || step != -1,
+        "snapped() called on vector component i32::MIN with step component -1"
+    );
```
Comment (Bromeon) [inline godot-core/src/builtin/vectors/vector_macros.rs:464]: Is only this a problem? `step == -1` would work?

**Comment by GodotRust** [comment]
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-768

File: godot-core/src/builtin/vectors/vector_macros.rs (hunk 4 above)
Comment (joriskleiber) [reply godot-core/src/builtin/vectors/vector_macros.rs:464]: No, I just realized how many overflow errors this function has... I have no idea how to handle all these, maybe you could give some guidance what the best way forward would be?

Comment (Bromeon) [reply godot-core/src/builtin/vectors/vector_macros.rs:464]: Limit the input range to the common usage (and panic for others) would be a start. Upon user request, we can then expand in the future.

**Comment by joriskleiber** [comment]
So I did the math (hopefully correct) and came up with these, in my opinion reasonable, panic conditions. I hope I didn't miss any other panics or overflows.

File: godot-core/src/builtin/vectors/vector_macros.rs [hunk 5]
```diff
@@ -457,6 +457,61 @@ macro_rules! impl_vector_fns {
...
+    assert!(
+        value != i32::MAX || step <= 1,
+        "snapped() called on vector component i32::MAX with step component greater then 1"
+    );
+
+    if step != 0 {
+        // Can overflow if step < -1 and value is i32::MIN or step > 1 and value is i32::MAX.
+        let a = value + step / 2;
+
+        // Manual implement `a.div_floor(step)` since Rust's native method is still unstable, as of 1.79.0.
```
Comment (Bromeon) [inline godot-core/src/builtin/vectors/vector_macros.rs:474]:
```suggestion
        // Manually implement `a.div_floor(step)` since Rust's native method is still unstable, as of 1.79.0.
```

File: godot-core/src/builtin/vectors/vector_macros.rs [hunk 6]
```diff
@@ -457,6 +457,61 @@ macro_rules! impl_vector_fns {
...
 
+pub(super) fn snap_one(mut value: i32, step: i32) -> i32 {
+    assert!(
+        value != i32::MIN || step >= 0,
+        "snapped() called on vector component i32::MIN with step component smaller then 0"
+    );
+    assert!(
+        value != i32::MAX || step <= 1,
+        "snapped() called on vector component i32::MAX with step component greater then 1"
+    );
```
Comment (Bromeon) [inline godot-core/src/builtin/vectors/vector_macros.rs:468]:
```suggestion
    assert!(
        value != i32::MIN || step >= 0,
        "snapped() called on vector component i32::MIN with step component less than 0"
    );
    assert!(
        value != i32::MAX || step <= 1,
        "snapped() called on vector component i32::MAX with step component greater than 1"
    );
```

File: godot-core/src/builtin/vectors/vector_macros.rs [hunk 7]
```diff
@@ -457,6 +457,61 @@ macro_rules! impl_vector_fns {
...
+    (
+        // Names of the components, for example `x, y`.
+        $($comp:ident),*
+    ) => {
+        /// A new vector with each component snapped to the closest multiple of the corresponding
+        /// component in `step`.
+        ///
+        /// # Panics
+        /// If any component of `self` is `i32::MIN` while the same component on `step` smaller then `0` or
+        /// if any component of `self` is `i32::MAX` while the same component on `step` greater then `1`.
```
Comment (Bromeon) [inline godot-core/src/builtin/vectors/vector_macros.rs:502]: Maybe this could be worded in a way makes clear that panics only happen in under/overflow scenarios, not general usage and especially not 0 steps:

```suggestion
        /// # Panics
        /// On under- or overflow:
        /// - If any component of `self` is `i32::MIN` while the same component on `step` less than `0`.
        /// - If any component of `self` is `i32::MAX` while the same component on `step` greater than `1`.
```

**Review by Bromeon — COMMENTED** [review]
Thanks a lot for approaching this so carefully! Turns out the seemingly simple `snapped` is very involved 😮

PR looks good, just some small doc nitpicks and it should be ready!

**Comment by joriskleiber** [comment]
As expected there where more possible overflow scenarios. For whatever reason i thought addition could only overflow for `i32::MIN - 1` or  `i32::MAX + 1`. 

The description for the second panic condition is horrific but I don't know how to word it otherwise.

**Review by Bromeon — APPROVED** [review]
Thanks for working out the conditions so carefully!

I think we might need to keep an eye on the performance; with all the overflow checks, it could happen that the happy path is now unnecessarily slow. But if this becomes a problem, we can probably use `debug_assert!` or similar.

**Comment by Ughuuu** [comment]
I second them becoming debug_asserts, but probably would need to be a bigger change over more things.

**Comment by Bromeon** [comment]
Thanks!
